master/buildbot/www/hooks/bitbucket.py

```diff
@@ -12,24 +12,170 @@
 # Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 #
 # Copyright Buildbot Team Members
-# Copyright 2013 (c) Mamba Team
-import json
+import logging
 
 from dateutil.parser import parse as dateparse
 
 from twisted.python import log
 
+try:
+    import json
+    assert json
+except ImportError:
+    import simplejson as json
 
-def getChanges(request, options=None):
-    """Catch a POST request from BitBucket and start a build process
+
+_HEADER_CT = 'Content-Type'
+_HEADER_EVENT = 'X-Event-Key'
+
+
+class BitbucketEventHandler(object):
+
+    def __init__(self, codebase=None):
+        self._codebase = codebase
+
+    def process(self, request):
+        """
+        Process a webhook event request.
+
+        See example on https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/hooks
+        for list of all the different events.
+        """
+        payload = self._get_payload(request)
+        event_type = request.getHeader(_HEADER_EVENT)
+        log.msg("Processing event %s: %r" % (_HEADER_EVENT, event_type,), logLevel=logging.DEBUG)
+        event_type = event_type.replace(":", "_")
+        handler = getattr(self, 'handle_%s' % event_type, None)
+
+        if handler is None:
+            raise ValueError('Unknown event: %r' % (event_type,))
+
+        return handler(payload)
+
+    def _get_payload(self, request):
+        content = request.content.read()
+        content_type = request.getHeader(_HEADER_CT)
+
+        if content_type == 'application/json':
+            payload = json.loads(content)
+        elif content_type == 'application/x-www-form-urlencoded':
+            payload = json.loads(request.args['payload'][0])
+        else:
+            raise ValueError('Unknown content type: %r' % (content_type,))
+
+        log.msg("Payload: %r" % payload, logLevel=logging.DEBUG)
+
+        return payload
+
+    def handle_repo_push(self, payload):
+        repo = payload['repository']
+        project = repo['name']
+        push = payload['push']
+        payload['ref'] = push['changes'][0]["new"]["name"]
+        repo_url = "git@bitbucket.org:%s.git" % payload['repository']['full_name']  # TOFIX: Hard coded as clone URL is not available
```

Comment:
But that won't fix the problem in the case where the clone URL is `https://<user>@bitbucket.org/<team>/<repo>.git`?